openvdb/openvdb/points/PointRasterizeFrustum.h

```diff
@@ -0,0 +1,1805 @@
+        if (clipBBox) {
+            frustumBBox.intersect(*clipBBox);
+        }
+
+        Vec3i outMin = frustumBBox.min().asVec3i();
+        Vec3i outMax = frustumBBox.max().asVec3i();
+
+        const bool interrupt = interrupter && frustumBBox.volume() > (32*32*32);
+
+        // back-project each output voxel into the input tree.
+        Vec3R xyz;
+        Coord outXYZ;
+        int &x = outXYZ.x(), &y = outXYZ.y(), &z = outXYZ.z();
+        for (x = outMin.x(); x <= outMax.x(); ++x) {
+            if (interrupt && interrupter->wasInterrupted()) break;
+            xyz.x() = x;
+            for (y = outMin.y(); y <= outMax.y(); ++y) {
+                if (interrupt && interrupter->wasInterrupted()) break;
+                xyz.y() = y;
+                for (z = outMin.z(); z <= outMax.z(); ++z) {
+                    xyz.z() = z;
+
+                    // approximate conversion to world-space
+
+                    Vec3d offset = (xyz - position) * voxelSize;
+
+                    double distanceSqr = offset[0]*offset[0]+offset[1]*offset[1]+offset[2]*offset[2];
+
+                    op(outXYZ, scale, attributeScale, distanceSqr, radiusWS*radiusWS);
+                }
+            }
+        }
+    }
+
+    template <typename SphereOpT>
+    static void rasterFrustumSphere(const Vec3d& position, const double scale,
+        const AttributeT& attributeScale, const float radiusWS,
+        const math::Transform& frustum, const CoordBBox* clipBBox,
+        util::NullInterrupter* interrupter, SphereOpT& op)
+    {
+        const Vec3d positionWS = frustum.indexToWorld(position);
+
+        BBoxd inputBBoxWS(positionWS-radiusWS, positionWS+radiusWS);
+        // Transform the corners of the input tree's bounding box
+        // and compute the enclosing bounding box in the output tree.
+        Vec3R frustumMin;
+        Vec3R frustumMax;
+        for (int i = 0; i < 8; ++i) {
+            Vec3R cornerWS(
+                i & 1 ? inputBBoxWS.max().x() : inputBBoxWS.min().x(),
+                i & 2 ? inputBBoxWS.max().y() : inputBBoxWS.min().y(),
+                i & 4 ? inputBBoxWS.max().z() : inputBBoxWS.min().z());
+            Vec3R corner = frustum.worldToIndex(cornerWS);
+            if (i == 0) {
+                 frustumMin = corner;
+                 frustumMax = corner;
+            } else {
+                frustumMin = math::minComponent(frustumMin, corner);
+                frustumMax = math::maxComponent(frustumMax, corner);
+            }
+        }
+        CoordBBox frustumBBox(Coord::floor(frustumMin), Coord::ceil(frustumMax));
+
+        // if clipping to frustum is enabled, clip the index bounding box
+        if (clipBBox) {
+            frustumBBox.intersect(*clipBBox);
+        }
+
+        Vec3i outMin = frustumBBox.min().asVec3i();
+        Vec3i outMax = frustumBBox.max().asVec3i();
+
+        const bool interrupt = interrupter && frustumBBox.volume() > (32*32*32);
```

Comment:
document constant (maybe make it a static int on the class with a descriptive name)